compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::span::Span;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind},
+    cooked::TokenKind,
+    Delim,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    pub span: Span,
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+/// Applies formatting rules to the given code str, generating edits where
+/// the source code needs to be changed to comply with the format rules.
+pub fn format(code: &str) -> Vec<Edit> {
+    let tokens = concrete::ConcreteTokenIterator::new(code);
+    let mut edits = vec![];
+
+    let mut indent_level = 0;
+
+    // The sliding window used is over three adjacent tokens
+    #[allow(unused_assignments)]
+    let mut one = None;
+    let mut two = None;
+    let mut three = None;
+
+    for token in tokens {
+        // Advance the token window
+        one = two;
+        two = three;
+        three = Some(token);
+
+        let mut edits_for_triple = match (&one, &two, &three) {
+            (Some(one), Some(two), Some(three)) => {
+                // if the token is a {, increase the indent level
+                if let ConcreteTokenKind::Syntax(TokenKind::Open(Delim::Brace)) = one.kind {
+                    indent_level += 1;
+                }
+
+                // if the token is a }, decrease the indent level
+                if let ConcreteTokenKind::Syntax(TokenKind::Close(Delim::Brace)) = one.kind {
+                    #[allow(clippy::implicit_saturating_sub)]
+                    if indent_level > 0 {
+                        indent_level -= 1;
+                    }
+                }
+
+                if let ConcreteTokenKind::WhiteSpace = one.kind {
+                    // first token is whitespace, continue scanning
+                    continue;
+                } else if let ConcreteTokenKind::WhiteSpace = two.kind {
+                    // whitespace in the middle
+                    apply_rules(
+                        one,
+                        get_token_contents(code, two),
+                        three,
+                        code,
+                        indent_level,
+                    )
+                } else {
+                    // one, two are adjacent tokens with no whitespace in the middle
+                    apply_rules(one, "", two, code, indent_level)
+                }
```

Comment:
I think this whole logic can be rewritten as:
```rust
match (one.kind, two.kind) {
  (Syntax(Open(Delim::Brace)), _) => ident_level += 1,
  (Syntax(Close(Delim::Brace)), _) => ident_level = ident_level.saturating_sub(1),
  (WhiteSpace, _) => (),
  (_, Whitespace) => apply_rules(one, get_token_contents(code, two), three, code, indent_level),
  _ => apply_rules(one, "", two, code, indent_level),
}
```
and in that way you can better express the intent that it depends on `one.kind` and `two.kind`.

Reply:
This doesn't allow the indentation to be adjusted in the same iteration as the rules applied, so it is a bit different.

Reply:
applied your suggestions to use the `match!()` syntax